lib/advisory_db.rb

```diff
@@ -0,0 +1,101 @@
+    version_advisory_files.each do |advisory|
+      File.open(advisory) do |_f|
+        advisory = YAML.load_file(advisory)
+        details << { title: advisory['title'], url: advisory['url'] } unless ignored_cves.include? advisory['cve']
+      end
+    end
+    details
+  end
+
+  def self.contain?(advisory, version)
+    advisory = YAML.load_file(advisory)
+    version_array = Array.new(1, version.number)
+    remove_unaffected_versions(version_array, (advisory['patched_versions'].to_a + advisory['unaffected_versions'].to_a).flatten).present?
+  end
+
+  def self.mark_versions
+    advisory = {}
+    count = 0
+    Dir.chdir(Rails.root.join('tmp', 'advisories', 'ruby-advisory-db'))
+    ActiveRecord::Base.transaction do
+      Dir.glob("gems/**/*.yml") do |advisory_file|
+        advisory = YAML.load_file(advisory_file)
+        rubygem = Rubygem.find_by(name: advisory['gem'])
+        unless rubygem.nil?
+          all_versions = rubygem.versions
+          version_number_list = all_versions.pluck(:number)
+          unaffected_versions = (advisory['patched_versions'].to_a + advisory['unaffected_versions'].to_a).flatten
+          vulnerable_versions = remove_unaffected_versions(version_number_list, unaffected_versions)
+          all_versions.where(number: vulnerable_versions).each do |vuln_version|
+            next if vuln_version.advisories.include? Advisory.where(cve: advisory['cve'])
+            vuln_version.advisories.create(url: advisory['url'], cve: advisory['cve'], title: advisory['title'],
+                                           description: advisory['description'])
+            count += 1
+          end
+        end
+      end
+    end
+    "Versions are successfully marked as vulnerable. #{count} new advisories marked."
+  end
+
+  def self.remove_unaffected_versions(version_number_list, unaffected_versions)
+    version_array = []
+    unaffected_versions.each do |versions|
+      version_array << if versions.split.length > 2
+                         versions.split(',')
+                       else
+                         versions
+                       end
+    end
+    version_array.flatten!
+    version_array.each do |versions|
+      unaffectedv_array = versions.split
+      unaffected_version_number = unaffectedv_array.second
+      case unaffectedv_array.first
+      when ">="
+        version_number_list.delete_if { |v| Gem::Version.new(v) >= Gem::Version.new(unaffected_version_number) }
+      when "<="
+        version_number_list.delete_if { |v| Gem::Version.new(v) <= Gem::Version.new(unaffected_version_number) }
+      when "~>"
+        version_number_list.delete_if { |v| Gem::Version.new(v) == Gem::Version.new(unaffected_version_number) }
+      when ">"
+        version_number_list.delete_if { |v| Gem::Version.new(v) > Gem::Version.new(unaffected_version_number) }
+      when "<"
+        version_number_list.delete_if { |v| Gem::Version.new(v) < Gem::Version.new(unaffected_version_number) }
```

Comment:
Maybe `Gem::Version.new(unaffected_version_number)` could be refactored into a local var.

Reply:
will do! 🙂